Simplenote/src/main/java/com/automattic/simplenote/viewmodels/TagsViewModel.kt

```diff
@@ -0,0 +1,120 @@
+package com.automattic.simplenote.viewmodels
+
+import android.view.View
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.automattic.simplenote.analytics.AnalyticsTracker
+import com.automattic.simplenote.models.TagItem
+import com.automattic.simplenote.repositories.TagsRepository
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+
+class TagsViewModel(private val tagsRepository: TagsRepository) : ViewModel() {
+    private val _uiState = MutableLiveData<UiState>()
+    val uiState: LiveData<UiState> = _uiState
+
+    private val _event = SingleLiveEvent<TagsEvent>()
+    val event: LiveData<TagsEvent> = _event
+
+    private var jobTagsFlow: Job? = null
+
+    fun start() {
+        viewModelScope.launch {
+            val tagItems = withContext(Dispatchers.IO) { tagsRepository.allTags() }
+            _uiState.value = UiState(tagItems)
+        }
+
+        jobTagsFlow = viewModelScope.launch {
+            tagsRepository.tagsChanged().collect {
+                val searchQuery = _uiState.value?.searchQuery
+                updateUiState(searchQuery)
+            }
+        }
+    }
+
+    private suspend fun updateUiState(searchQuery: String?, searchUpdate: Boolean = false) {
+        val tagItems = withContext(Dispatchers.IO) {
+            if (searchQuery == null) tagsRepository.allTags()
+            else tagsRepository.searchTags(searchQuery)
+        }
+        _uiState.value = UiState(tagItems, searchUpdate, searchQuery)
+    }
+
+    fun clickAddTag() {
+        _event.postValue(TagsEvent.AddTagEvent)
+    }
+
+    fun longClickAddTag() {
+        _event.postValue(TagsEvent.LongAddTagEvent)
+    }
+
+    fun close() {
+        _event.postValue(TagsEvent.FinishEvent)
+    }
+
+    fun closeSearch() {
+        _uiState.value = _uiState.value?.copy(searchUpdate = true, searchQuery = null)
+    }
+
+    fun search(searchQuery: String) {
+        viewModelScope.launch {
+            updateUiState(searchQuery, true)
+        }
+    }
+
+    fun pause() {
```

Comment:
> Consider adding tests for this function.

Use a similar test to the test for `collect { ... }` logic with the difference that when we call `pause`, any changes to the tags is not listened, therefore the `UiState` is not updated. Fixed in 79c13d4b073b2f1b200967ea0ee9e67b183ffe63.

Reply:
> Do you think this comment is necessary?

Since `callbackFlow` is a new library and it is not explicit and clear that `cancel()` eventually calls to `awaitClose()`, I think the comment is useful for any reader which has not used `callbackFlow` before.

Reply:
👋 @dandominguez04 !

> Fixed in 79c13d4.

Thanks for the fix! 🙏 

> Since callbackFlow is a new library and it is not explicit and clear that cancel() eventually calls to awaitClose(), I think the comment is useful for any reader which has not used callbackFlow before.

This makes sense, let's keep it then! 👍 Maybe, one last suggestion then would be to move this comment(s) to a Javadoc on top of the `pause()` function itself. But, it is totally up to you at this point.